Add transaction error tracking section (#24)
snowplow/snowplow-ecommerce-tracking-accelerator: content/tracking/transaction_error.md

+++
title = "Track Transaction Errors"
weight = 8
post = ""
+++

> Transaction error events can be used to track issues during any step of a transaction being processed by the system. Even if the API is generic and can be used in any case of a transaction not being completed, there is a plethora of payment processor/gateway errors which are extremely useful to include in your warehouse together with the rich Snowplow schema such as:

Reviewer: ```suggestion
Transaction error events can be used to track issues during any step of a transaction being processed by the system. The API is generic and can be used in any case of an incomplete transaction, including payment processor/gateway errors. These data can be combined with the rich Snowplow schema to provide insights such as:
```
Slight wording change

- Fraud rejections and other risk related rejections
- Processor/gateway internal errors

> Additionally by tracking the error types, soft or hard and the resolution selected by your system, you can test and optimize the flows and experience you provide to your customers facing issues with completing a transaction.

Reviewer: ```suggestion
Additionally by tracking the error types, soft or hard, and the resolution selected by your system, you can test and optimize the flows and experience you provide to your customers facing issues with completing a transaction.
```
Comma

- Where `error_code` is the error-identifying code for the transaction issue E.g. E522.
- Where `error_shortcode` is the shortcode for the transaction error E.g declined_by_api, pm_card_radarBlock.
- Where `error_description` is the longer description for the transaction error.
- Where `error_type` is the type of error the specific issue is regarded as in the standard classification of _hard_ and _soft_. Hard error types mean the customer must provide another form of payment e.g. an expired card. Soft errors can be the result of temporary issues where retrying might be successful e.g. processor declined the transaction.
- Where `transaction` is the transaction the issue occurred in.

**Example usage:**

```ts
window.snowplow("trackTransactionError:{trackerName}", {
  resolution: "rejection",
  error_code: "E123",
  error_shortcode: "CARD_DECLINE",
  error_description: "Card has been declined by the issuing bank.",
  error_type: "hard",
  transaction: {
    revenue: 45,
    currency: "EUR",
    transaction_id: "T12345",
    payment_method: "card" 
  }
});
```

{{% /tab %}}

{{% tab name="Node.js" %}}
#### Track transaction errors using a self describing event

> To track transaction errors in the Node.js tracker, you can use the `buildSelfDescribingEvents` method together with the correct parameters for the transaction error event.

Reviewer: ```suggestion
To track transaction errors in the Node.js tracker, you can use the `buildSelfDescribingEvent` method together with the correct parameters for the transaction error event.
```
I think the method is `buildSelfDescribingEvent`, if not it's wrong in the code snippet

Author: Yeap all fixed, thanks!

For the transaction taking part in the process, an additional transaction data context is added on the contexts to be sent with the event.

**Example usage:**

```ts
/* Where `t` is a placeholder for the initialized tracker instance on your application.  */

t.track(buildSelfDescribingEvent({
  event: {
    schema: "iglu:com.snowplowanalytics.snowplow.ecommerce/snowplow_ecommerce_action/jsonschema/1-0-2",
    data: {
      type: "trns_error"
    }
  }
}, [
  /* Transaction Error context */
  {
    schema: "iglu:com.snowplowanalytics.snowplow.ecommerce/transaction_error/jsonschema/1-0-0",
    data: { 
      resolution: "rejection",
      error_code: "E123",
      error_shortcode: "CARD_DECLINE",
      error_description: "Card has been declined by the issuing bank.",
      error_type: "hard",
    }
  },
  /* Transaction context */
  {
    schema: "iglu:com.snowplowanalytics.snowplow.ecommerce/transaction/jsonschema/1-0-0",
    data: {
      revenue: 45,
      currency: "EUR",
      transaction_id: "T12345",
      payment_method: "card" 
    }
  }
]));
```

> To learn more about tracking self describing events in Node.js take a look at the `buildSelfDescribingEvents` [documentation](https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/javascript-trackers/node-js-tracker/node-js-tracker-v3/tracking-events/#track-self-describing-events-withbuildselfdescribingevent).

Reviewer: ```suggestion
To learn more about tracking self describing events in Node.js take a look at the `buildSelfDescribingEvent` [documentation](https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/javascript-trackers/node-js-tracker/node-js-tracker-v3/tracking-events/#track-self-describing-events-withbuildselfdescribingevent).
```
